File: vivisector/report.py
#/usr/bin/env python

import os
import urllib


class ReportWriter(object):

    def __init__(self):
        self.titlePrefix = "Vivisector report for"
        self.masterCssClass = "masterimg"
        self.holderCssClass = "holder"
        self.extraCss = ""

    def filename_of_report(self, report_name):
        return "%s.html" % report_name

    def process(self, report_name, matched_sets):
        filename = self.filename_of_report(report_name)
        print filename
        title = "%s %s" % (self.titlePrefix, report_name)
        out_html = ""
        out_html += "<html>\n<head>\n"
        out_html += "<title>%s</title>\n" % title
        out_html += "<style type='text/css'>\n"

        out_html += "body {color:white; background-color:#333;}\n"
        out_html += "th {border-top: 1px solid #ccc;}\n"
        out_html += "td {padding-bottom:1ex;}\n"
        out_html += ".masterimg {background-color:white;}\n"
        out_html += ".missing {white-space: pre; text-align:center;}\n"

        out_html += self.extraCss

        out_html += "</style>\n"
        out_html += "</head>\n"
        out_html += "<body>\n"

        out_html += "<h1>%s</h1>\n" % title

        if 0 == len(matched_sets):
            out_html += "No input images were found"
        else:
            out_html += "<table class='comparison'>\n"
            out_html += " <tr>\n"
            for (_, target, __) in matched_sets[0]:
                out_html += "  <td align='center'>%s</td>\n" % target
            out_html += " </tr>\n"

            for row in matched_sets:
                (_, __, img_master) = row[0]
                img_actual = row[1:]
                out_html += " <tr>\n"
                base = os.path.basename(img_master)
                out_html += "  <th colspan='%d'><a name='%s'>%s</a></th>\n" % (len(row), urllib.quote(base), base)
                out_html += " </tr>\n"
                out_html += " <tr>\n"
                out_html += "  <td align='right' valign='top'>"
                out_html +=    "<img class='masterimg' src='file://%s' alt='%s' /></td>\n" % (img_master, base)

                for (design_image, target, img) in img_actual:
                    if img is None:
                        out_html += "  <td><div class='missing'>%s\non %s</div></td>\n" % (design_image, target)
                    else:
                        out_html += "  <td align='left' valign='top'>"
                        out_html +=    "<div class='%s'><img src='file://%s' class='ipad' alt='%s' /></div></td>\n" % (self.holderCssClass, img, img)
                out_html += " </tr>\n"
            out_html += "</table>"

        out_html += "</body></html>"

        with open(filename, "w") as text_file:
            text_file.write(out_html)

File: gem/lib/vivisector/report.rb
require 'markaby'
require_relative 'filesystem'

module Vivisector

  class Report
    attr_accessor :title
    attr_accessor :attribute
    attr_accessor :anatomy
    attr_accessor :appographies
    attr_accessor :web_document_root
    attr_accessor :destination_path
    attr_accessor :img_width_px

    def css
      width_string = ""
      width_string = "width: #{@img_width_px}px;" unless img_width_px.nil?
      %(
        body {color:white; background-color:#333;}
        table.comparison th {border-top: 1px solid #ccc;}
        table.comparison td {padding-bottom:1ex; text-align:center;}
        .masterimg, .appimg {background-color:white; #{width_string}}
        .img {border:0;}
        .missing {text-align:center; #{width_string}}
       )
    end

    # if necessary, modify the path to be relative (for web-based reports)
    def path_transform element_path
      unless @web_document_root.nil?
        element_path = web_relative_path(@web_document_root, @destination_path, element_path)
      end
      element_path
    end

    def to_s
      # initial calculations
      candidates = @appographies.select do |appog|
        next true if @attribute.nil? # unless there is nothing in this candidate???? might be expensive to check.

        appog.image_set.target.attribute == @attribute
      end
      design_images = @anatomy.image_set.images_for_attribute(@attribute)
      columns = candidates.length + 1

      me = self

      # build html
      mab = Markaby::Builder.new
      mab.html do

        head do
          title "#{me.title} | Vivisector"
          style :type => "text/css" do
            me.css
          end
        end

        body do
          h1 me.title
          if design_images.empty?
            p "No input images were found."
          else
            table.comparison do
              # print out the first row of the table -- the target names
              tr do
                td "Design"
                candidates.each do |c|
                  td c.image_set.target.name
                end
              end

              # print out all the compared images
              design_images.each do |design_image|
                # title
                tr do
                  th :colspan => columns do
                    a :name => design_image.description do
                      design_image.description
                    end
                  end
                end

                # images
                tr do
                  td :align => "right", :valign => "top" do
                    src = me.path_transform(design_image.path)
                    a :href => src do
                      img.masterimg :src => src, :alt => design_image.description
                    end
                  end

                  candidates.each do |candidate|
                    app_image = candidate.image_set.best_image_for(design_image.id)
                    if app_image.nil?
                      td { div.missing "#{design_image.description} on #{candidate.image_set.target.name}" }
                    else
                      td :align => "left", :valign => "top" do
                        div.holder do
                          src = me.path_transform(app_image.path)
                          a :href => src do
                            img.appimg :src => src, :alt => app_image.description
                          end
                        end
                      end
                    end
                  end
                end
              end
            end
          end
        end
      end
      return mab.to_s
    end

  end #Report

end
